cmake/onnxruntime_external_deps.cmake

```diff
@@ -0,0 +1,408 @@
+# External dependencies
+if (onnxruntime_BUILD_BENCHMARKS)
+  # We will not need to test benchmark lib itself.
+  set(BENCHMARK_ENABLE_TESTING OFF CACHE BOOL "Disable benchmark testing as we don't need it.")
+  # We will not need to install benchmark since we link it statically.
+  set(BENCHMARK_ENABLE_INSTALL OFF CACHE BOOL "Disable benchmark install to avoid overwriting vendor install.")
+
+  FetchContent_Declare(
+    google_benchmark
+    URL ${DEP_URL_google_benchmark}
+    URL_HASH SHA1=${DEP_SHA1_google_benchmark}
+  )
+endif()
+
+if (NOT WIN32)
+    FetchContent_Declare(
+    google_nsync
+    URL ${DEP_URL_google_nsync}
+    URL_HASH SHA1=${DEP_SHA1_google_nsync}
+    FIND_PACKAGE_ARGS NAMES nsync
+    )
+endif()
+list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/external)
+
+# ABSL should be included before protobuf because protobuf may use absl
+if(NOT onnxruntime_DISABLE_ABSEIL)
+  include(external/abseil-cpp.cmake)
+endif()
+
+# Flatbuffers
+# We do not need to build flatc for iOS or Android Cross Compile
+if (CMAKE_SYSTEM_NAME STREQUAL "iOS" OR CMAKE_SYSTEM_NAME STREQUAL "Android" OR onnxruntime_BUILD_WEBASSEMBLY)
+  set(FLATBUFFERS_BUILD_FLATC OFF CACHE BOOL "FLATBUFFERS_BUILD_FLATC" FORCE)
+endif()
+set(FLATBUFFERS_BUILD_TESTS OFF CACHE BOOL "FLATBUFFERS_BUILD_TESTS" FORCE)
+set(FLATBUFFERS_INSTALL OFF CACHE BOOL "FLATBUFFERS_INSTALL" FORCE)
+set(FLATBUFFERS_BUILD_FLATHASH OFF CACHE BOOL "FLATBUFFERS_BUILD_FLATHASH" FORCE)
+set(FLATBUFFERS_BUILD_FLATLIB ON CACHE BOOL "FLATBUFFERS_BUILD_FLATLIB" FORCE)
+FetchContent_Declare(
+    flatbuffers
+    URL ${DEP_URL_flatbuffers}
+    URL_HASH SHA1=${DEP_SHA1_flatbuffers}
+    FIND_PACKAGE_ARGS NAMES Flatbuffers
+)
+
+#Here we support two build mode:
+#1. if ONNX_CUSTOM_PROTOC_EXECUTABLE is set, build Protobuf from source, except protoc.exe. This mode is mainly
+#   for cross-compiling
+#2. if ONNX_CUSTOM_PROTOC_EXECUTABLE is not set, Compile everything(including protoc) from source code.
+FetchContent_Declare(
+  Protobuf
+  URL ${DEP_URL_protobuf}
+  URL_HASH SHA1=${DEP_SHA1_protobuf}
+  SOURCE_SUBDIR  cmake
+  FIND_PACKAGE_ARGS NAMES Protobuf
+)
+
+if (CMAKE_SYSTEM_NAME STREQUAL "Android")
+  set(protobuf_BUILD_PROTOC_BINARIES OFF CACHE BOOL "Build protobuf tests" FORCE)
+  set(protobuf_WITH_ZLIB OFF CACHE BOOL "Build with zlib support" FORCE)
+endif()
+if (onnxruntime_DISABLE_RTTI)
+  set(protobuf_DISABLE_RTTI ON CACHE BOOL "Remove runtime type information in the binaries" FORCE)
+endif()
+
+include(protobuf_function.cmake)
+#protobuf end
+
+if (onnxruntime_DISABLE_CONTRIB_OPS)
+  add_definitions(-DDISABLE_CONTRIB_OPS)
+endif()
+
+if (onnxruntime_DISABLE_ML_OPS)
+  add_definitions(-DDISABLE_ML_OPS)
+endif()
+
+if (onnxruntime_DISABLE_SPARSE_TENSORS)
+  add_compile_definitions(DISABLE_SPARSE_TENSORS)
+endif()
+
+if (onnxruntime_DISABLE_OPTIONAL_TYPE)
+  add_compile_definitions(DISABLE_OPTIONAL_TYPE)
+endif()
```

Comment:
Do those qualify as external deps? If needed here, a comment would be helpful.

Reply:
Moved to the main CMakeLists.txt